components/eamxx/src/dynamics/homme/eamxx_homme_iop.cpp

```diff
@@ -341,49 +322,38 @@ apply_iop_forcing(const Real dt)
+      const int igp = (team.league_rank()/NGP)%NGP;
+      const int jgp =  team.league_rank()%NGP;
 
       // Get temp views from workspace
       auto ws = eamxx_wsm.get_workspace(team);
-      uview_1d<Pack> pmid, pint, pdel;
+      uview_1d<Pack> pmid;
       ws.take_many_contiguous_unsafe<1>({"pmid"},{&pmid});
 
-      Kokkos::parallel_for(Kokkos::TeamThreadRange(kv.team, NGP*NGP), [&] (const int idx) {
-        const int igp = idx/NGP;
-        const int jgp = idx%NGP;
-
-        auto ps_i          = ps_dyn(ie, igp, jgp);
-        auto dp3d_i        = ekat::subview(dp3d_dyn, ie, igp, jgp);
-        auto vtheta_dp_i   = ekat::subview(vtheta_dp_dyn, ie, igp, jgp);
-        auto qv_i          = ekat::subview(Q_dyn, ie, 0, igp, jgp);
-        auto temperature_i = ekat::subview(temperature, ie, igp, jgp);
-
-        // Compute reference pressures and layer thickness.
-        // TODO: Allow geometry data to allocate packsize
-        auto s_pmid = ekat::scalarize(pmid);
-        Kokkos::parallel_for(Kokkos::TeamVectorRange(team, total_levels), [&](const int& k) {
-          s_pmid(k) = hyam(k)*ps0 + hybm(k)*ps_i;
-        });
-        team.team_barrier();
-
-        // Reinterperate into views of Homme::Scalar for calling Hommexx function.
-        Homme::ExecViewUnmanaged<Homme::Scalar[NLEV]> dp3d_scalar(reinterpret_cast<Homme::Scalar*>(dp3d_i.data()), NLEV);
-        Homme::ExecViewUnmanaged<Homme::Scalar[NLEV]> vtheta_dp_scalar(reinterpret_cast<Homme::Scalar*>(vtheta_dp_i.data()), NLEV);
-        Homme::ExecViewUnmanaged<Homme::Scalar[NLEV]> qv_scalar(reinterpret_cast<Homme::Scalar*>(qv_i.data()), NLEV);
-        Homme::ExecViewUnmanaged<Homme::Scalar[NLEV]> temperature_scalar(reinterpret_cast<Homme::Scalar*>(temperature_i.data()), NLEV);
-
-        // Compute temperature from virtual potential temperature
-        Kokkos::parallel_for(Kokkos::ThreadVectorRange(team, NLEV), [&] (const int k) {
-          auto T_val = vtheta_dp_i(k);
-          T_val /= dp3d_i(k);
-          T_val = PF::calculate_temperature_from_virtual_temperature(T_val,qv_i(k));
-          temperature_i(k) = PF::calculate_T_from_theta(T_val,pmid(k));
-        });
+      auto ps_i          = ps_dyn(ie, igp, jgp);
+      auto dp3d_i        = ekat::subview(dp3d_dyn, ie, igp, jgp);
+      auto vtheta_dp_i   = ekat::subview(vtheta_dp_dyn, ie, igp, jgp);
+      auto qv_i          = ekat::subview(Q_dyn, ie, 0, igp, jgp);
+      auto temperature_i = ekat::subview(temperature, ie, igp, jgp);
+
+      // Compute reference pressures and layer thickness.
+      // TODO: Allow geometry data to allocate packsize
+      auto s_pmid = ekat::scalarize(pmid);
+      Kokkos::parallel_for(Kokkos::TeamVectorRange(team, total_levels), [&](const int& k) {
+        s_pmid(k) = hyam(k)*ps0 + hybm(k)*ps_i;
+      });
+      team.team_barrier();
 
+      // Compute temperature from virtual potential temperature
+      Kokkos::parallel_for(Kokkos::ThreadVectorRange(team, NLEV), [&] (const int k) {
```

Comment:
This now should be a `TeamVectorRange` since this is the 1st level of ||ism inside `TeamPolicy` loop, and there are no more interior loops to ||ize. You could also use a `TeamThreadRange`, but that could potentially leave some ||ism on the table. 

But `ThreadVectorRange` is assuming that we are inside a `TeamThreadRange` which we no longer are. So this is a bug.

Reply:
For clarity: you _can_ use TeamThreadRange, and you don't leave any ||ism on the table _as long as the policy was created with vector length 1_. That's usually the case for policies created in eamxx (homme's policies for instance are not like that, since they are tailored for 3 layers of ||ism).

If the policy has vector length>1, not only TeamThreadRange is wasting threads, but it may give bad answers, since all threads in the warp will run the same code (so stuff like += will be executed many times over).